proto/structured.proto

```diff
@@ -104,3 +105,224 @@ message CreateTableResponse {
+  optional TableKey key = 2 [(gogoproto.nullable) = false];
+  // get specified columns from the row.
+  repeated string column_names = 3;
+}
+
+// A GetTableRowResponse returns a row.
+// If the key doesn't exist, it returns an empty row.
+message GetTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableRow row = 2 [(gogoproto.nullable) = false];
+}
+
+// A PutTableRowRequest inserts/updates a row in the table. To write
+// an empty value in a cell specify an empty Bytes in the Value. To Delete
+// a cell, specify a cell with a column_name with no value.
+message PutTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableKey key = 2 [(gogoproto.nullable) = false];
+  optional TableRow row = 3 [(gogoproto.nullable) = false];
+}
+
+// A PutTableRowResponse is the response.
+message PutTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
+// A ConditionalPutTableRowRequest.
+//
+// - Returns true and sets row if exp_row equals existing row.
+// - If key doesn't exist and exp_row is empty, sets row.
+// - If key exists and exp_row is empty, sets row.
+// - Otherwise, returns error and the actual value of the row in the response.
+message ConditionalPutTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableKey key = 2 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // The values to put.
+  optional TableRow row = 3 [(gogoproto.nullable) = false];
+  optional TableRow exp_row = 4[(gogoproto.nullable) = false];
+}
+
+// A ConditionalPutTableRowResponse is the return value from the
+// ConditionalPut request.
+message ConditionalPutTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableRow row = 2[(gogoproto.nullable) = false];
+}
+
+// A DeleteTableRowRequest deletes the entire row.
+message DeleteTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableKey key = 2 [(gogoproto.nullable) = false];
+}
+
+// A DeleteTableRowResponse.
+message DeleteTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
+// A DeleteTableRowRangeRequest deletes the specified rows.
+message DeleteTableRowRangeRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // Delete *all* entries between Key (inclusive) and EndKey
+  // (exclusive). Must be > 0
+  optional string index = 2 [(gogoproto.nullable) = false];
+  optional TableKey key = 3 [(gogoproto.nullable) = false];
+  optional TableKey end_key = 4 [(gogoproto.nullable) = false];
```

Comment:
With `table_name` inside of `TableKey` you have it duplicated in this request and `ScanTableRequest`. The duplication is of little concern except that it opens up the possibility for someone to specify them differently. Like `index`, I think `table_name` belongs at the request level.